src/target/riscv/riscv-013_reg.c

```diff
@@ -13,73 +13,452 @@
+static int fetch_vreg(struct reg *reg)
+{
+	return riscv013_get_register_buf(riscv_reg_impl_get_target(reg),
+			reg->value, reg->number);
+}
 
-		reg->valid = riscv_reg_impl_gdb_regno_cacheable(reg->number, /* is write? */ false);
-	} else {
-		uint64_t value;
-		int result = riscv_reg_get(target, &value, reg->number);
-		if (result != ERROR_OK)
-			return result;
-		buf_set_u64(reg->value, 0, reg->size, value);
+static int send_vreg(const struct reg *reg, const uint8_t *buf)
+{
+	return riscv013_set_register_buf(riscv_reg_impl_get_target(reg),
+			reg->number, reg->value);
+}
+
+/* End of register fetch/send definitions. */
+/* Start of cache entry utils. */
+
+static void set_cache_value(struct reg *reg, const uint8_t *buf)
+{
+	assert(reg);
+	assert(reg->value);
+	assert(reg->size);
+	buf_cpy(buf, reg->value, reg->size);
+}
+
+/**
+ * Note: there are only 3 states a cahce entry can be in:
+ * |: state   :|: "reg->valid" :|: "reg->dirty" :|
+ * | "invalid" | false          | false          |
+ * | "valid"   | true           | false          |
+ * | "dirty"   | true           | true           |
+ *
+ * The state "reg->valid == dirty && reg->valid == false" should be unreacheable.
+ * To achieve this, "reg->valid" and "reg->dirty" are not assigned directly
+ * outside of "mark_<state>()" functions.
+ */
+static void mark_invalid(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	reg->valid = false;
+	reg->dirty = false;
+}
+
+static void mark_clean(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	assert(riscv_reg_impl_get_target(reg)->state == TARGET_HALTED);
+	reg->valid = true;
+	reg->dirty = false;
+}
+
+static void mark_dirty(struct reg *reg)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
+	assert(riscv_reg_impl_get_target(reg)->state == TARGET_HALTED);
+	reg->valid = true;
+	reg->dirty = true;
+}
+
+static int access_prelude(const struct reg *reg, bool write)
+{
+	assert(riscv_reg_impl_is_initialized(reg));
+	const struct target * const target = riscv_reg_impl_get_target(reg);
+	if (!reg->exist) {
+		LOG_TARGET_DEBUG(target, "Register %s does not exist.",
+				reg->name);
+		return ERROR_FAIL;
 	}
-	char *str = buf_to_hex_str(reg->value, reg->size);
-	LOG_TARGET_DEBUG(target, "Read 0x%s from %s (valid=%d).", str, reg->name,
-			reg->valid);
-	free(str);
+	LOG_TARGET_DEBUG(target, "%s %s (valid=%s, dirty=%s).",
+			write ? "Writing" : "Reading", reg->name,
+			reg->valid ? "true" : "false",
+			reg->dirty ? "true" : "false");
 	return ERROR_OK;
 }
 
-static int riscv013_reg_set(struct reg *reg, uint8_t *buf)
+static void access_epilogue(const struct reg *reg, bool write)
 {
-	struct target *target = riscv_reg_impl_get_target(reg);
-
-	char *str = buf_to_hex_str(buf, reg->size);
-	LOG_TARGET_DEBUG(target, "Write 0x%s to %s (valid=%d).", str, reg->name,
-			reg->valid);
+	assert(riscv_reg_impl_is_initialized(reg));
+	assert(reg->exist);
```

Comment:
The predicate `riscv_reg_impl_is_existing()`, suggested in another comment, could perhaps be used here, too:

```suggestion
	assert(riscv_reg_impl_is_existing(reg));
```

Reply:
Addressed